In [1]:
import pandas as pd
import matplotlib

In [2]:
df = pd.read_csv("../data/kaggle_survey_2020_responses.csv", low_memory=False)

In [3]:
df.head(2)

Time from Start to Finish (seconds)                           Q1  \
0               Duration (in seconds)  What is your age (# years)?   
1                                1838                        35-39   

                                       Q2  \
0  What is your gender? - Selected Choice   
1                                     Man   

                                          Q3  \
0  In which country do you currently reside?   
1                                   Colombia   

                                                  Q4  \
0  What is the highest level of formal education ...   
1                                    Doctoral degree   

                                                  Q5  \
0  Select the title most similar to your current ...   
1                                            Student   

                                                  Q6  \
0  For how many years have you been writing code ...   
1                                         5-10 years   

                                           Q7_Part_1  \
0  What programming languages do you use on a reg...   
1                                             Python   

                                           Q7_Part_2  \
0  What programming languages do you use on a reg...   
1                                                  R   

                                           Q7_Part_3  ...  \
0  What programming languages do you use on a reg...  ...   
1                                                SQL  ...   

                                        Q35_B_Part_2  \
0  In the next 2 years, do you hope to become mor...   
1                                                NaN   

                                        Q35_B_Part_3  \
0  In the next 2 years, do you hope to become mor...   
1                                                NaN   

                                        Q35_B_Part_4  \
0  In the next 2 years, do you hope to become mor...   
1                                                NaN   

                                        Q35_B_Part_5  \
0  In the next 2 years, do you hope to become mor...   
1                                       TensorBoard    

                                        Q35_B_Part_6  \
0  In the next 2 years, do you hope to become mor...   
1                                                NaN   

                                        Q35_B_Part_7  \
0  In the next 2 years, do you hope to become mor...   
1                                                NaN   

                                        Q35_B_Part_8  \
0  In the next 2 years, do you hope to become mor...   
1                                                NaN   

                                        Q35_B_Part_9  \
0  In the next 2 years, do you hope to become mor...   
1                                                NaN   

                                       Q35_B_Part_10  \
0  In the next 2 years, do you hope to become mor...   
1                                                NaN   

                                         Q35_B_OTHER  
0  In the next 2 years, do you hope to become mor...  
1                                                NaN  

[2 rows x 355 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20037 entries, 0 to 20036
Columns: 355 entries, Time from Start to Finish (seconds) to Q35_B_OTHER
dtypes: object(355)
memory usage: 54.3+ MB


In [5]:
df.size, df.shape

(7113135, (20037, 355))

In [6]:
df.describe()

Time from Start to Finish (seconds)     Q1     Q2     Q3  \
count                                20037  20037  20037  20037   
unique                                4196     12      6     56   
top                                    565  25-29    Man  India   
freq                                    36   4011  15789   5851   

                     Q4       Q5         Q6 Q7_Part_1 Q7_Part_2 Q7_Part_3  \
count             19570    19278      19121     15531      4278      7536   
unique                8       14          8         2         2         2   
top     Master’s degree  Student  3-5 years    Python         R       SQL   
freq               7859     5171       4546     15530      4277      7535   

        ...        Q35_B_Part_2 Q35_B_Part_3          Q35_B_Part_4  \
count   ...                1178          495                   431   
unique  ...                   2            2                     2   
top     ...   Weights & Biases     Comet.ml    Sacred + Omniboard    
freq    ...                1177          494                   430   

         Q35_B_Part_5 Q35_B_Part_6 Q35_B_Part_7 Q35_B_Part_8  \
count            3200          558          481          847   
unique              2            2            2            2   
top      TensorBoard     Guild.ai     Polyaxon       Trains    
freq             3199          557          480          846   

                  Q35_B_Part_9 Q35_B_Part_10 Q35_B_OTHER  
count                      520          3083         252  
unique                       2             2           2  
top      Domino Model Monitor           None       Other  
freq                       519          3082         251  

[4 rows x 355 columns]

## Examine Age data. Riddle number one: Is there a different age distribution in the survey than the csv data tell us ?

In [7]:
df['Q1'].unique()

array(['What is your age (# years)?', '35-39', '30-34', '22-24', '25-29',
       '18-21', '55-59', '50-54', '40-44', '60-69', '45-49', '70+'],
      dtype=object)

In [36]:
# df['Q1'].sort_index() # type(df['Q1']) # df['Q1'].dtypes

In [28]:
df['Q1'].value_counts(sort=True)  # default See: pd.value_counts?

25-29                          4011
22-24                          3786
18-21                          3469
30-34                          2811
35-39                          1991
40-44                          1397
45-49                           988
50-54                           698
55-59                           411
60-69                           398
70+                              76
What is your age (# years)?       1
Name: Q1, dtype: int64

In [24]:
df['Q1'].value_counts().sort_index()

18-21                          3469
22-24                          3786
25-29                          4011
30-34                          2811
35-39                          1991
40-44                          1397
45-49                           988
50-54                           698
55-59                           411
60-69                           398
70+                              76
What is your age (# years)?       1
Name: Q1, dtype: int64

In [38]:
df['Q1'].value_counts(normalize=True).sort_index()

18-21                          0.173130
22-24                          0.188950
25-29                          0.200180
30-34                          0.140290
35-39                          0.099366
40-44                          0.069721
45-49                          0.049309
50-54                          0.034836
55-59                          0.020512
60-69                          0.019863
70+                            0.003793
What is your age (# years)?    0.000050
Name: Q1, dtype: float64

In [ ]:
df['Q1'].value_counts

First remarks on Age:
Age groups could perhaps be left for competitors to construct.
Data could be presented as answered. It is interesting to see the large number of the 20-21 class, but some useful information may be lost.

Even more problematic:
There are 7 bins (25-29 until 55-59) that contain 5year-periods, one bin 60-69 that includes 10 years and 70+. While it is interesting that there are 76 responses 70 or more year olds, we do not have the max, which I find disappointing because you know, we all need some inspiration!
The real interesting part though is why the bin 18-21 includes 4 years and bin 22-24 includes only 3 ?

What would be the age distribution be if we had an equal length bin 20-24 and a bin under 20 ?
Let us explore this idea:
First of all, it is easy to show that the largest group would be 20 to 24 by a wide margin.
And this is a very interesting result on which all other info has to be conditioned upon.
First I will show that under all assumptions this assumption about the 20-25 age stands.

If 18-21 are evenly  distributed: => more or less 1156 per year => 2300 20 and 21 => age group: 20-24 = 3786 + 2300 = 6000 !

In [11]:
df_before_30 = df[(df['Q1'] == '18-21') |
   (df['Q1'] == '22-24') | 
   (df['Q1'] == '25-29')
  ]

In [12]:
df_before_30.shape  # more than half 

(11266, 355)

In [ ]:
df_22_24 = df[(df['Q1'] == '22-24')]

In [39]:
df_22_24 = df[(df['Q1'] == '22-24')]

In [13]:
df_18_21 = df[(df['Q1'] == '18-21')]

In [14]:
df_18_21.head(2)

Time from Start to Finish (seconds)     Q1     Q2     Q3  \
11                                7469  18-21    Man  India   
16                                 413  18-21  Woman  India   

                   Q4       Q5         Q6 Q7_Part_1 Q7_Part_2 Q7_Part_3  ...  \
11  Bachelor’s degree  Student  1-2 years    Python         R       SQL  ...   
16  Bachelor’s degree  Student  1-2 years       NaN         R       SQL  ...   

   Q35_B_Part_2 Q35_B_Part_3 Q35_B_Part_4 Q35_B_Part_5 Q35_B_Part_6  \
11          NaN          NaN          NaN          NaN          NaN   
16          NaN          NaN          NaN          NaN          NaN   

   Q35_B_Part_7 Q35_B_Part_8 Q35_B_Part_9 Q35_B_Part_10 Q35_B_OTHER  
11          NaN          NaN          NaN           NaN         NaN  
16          NaN          NaN          NaN          None         NaN  

[2 rows x 355 columns]

### Ideas on how to separate 18-21 to 18-19 (below 20) and 20-21

In [45]:
df_student_22_24 =  df_22_24[df_22_24['Q5'] == 'Student']
df_student_22_24.shape

(1479, 355)

In [47]:
df_22_24['Q5'].value_counts()

Student                      1479
Software Engineer             408
Currently not employed        401
Data Scientist                367
Data Analyst                  282
Machine Learning Engineer     228
Other                         158
Business Analyst               89
Research Scientist             75
Data Engineer                  53
Statistician                   53
Product/Project Manager        25
DBA/Database Engineer           8
Name: Q5, dtype: int64

## Riddle number two: What is the educational level in the survey?

In [15]:
df_18_21['Q4'].value_counts()

Bachelor’s degree                                                    2288
Some college/university study without earning a bachelor’s degree     451
Master’s degree                                                       414
I prefer not to answer                                                 76
No formal education past high school                                   66
Professional degree                                                    38
Doctoral degree                                                        18
Name: Q4, dtype: int64

In [41]:
df_22_24['Q4'].value_counts()

Bachelor’s degree                                                    1702
Master’s degree                                                      1554
Some college/university study without earning a bachelor’s degree     180
Professional degree                                                    90
Doctoral degree                                                        74
I prefer not to answer                                                 61
No formal education past high school                                   21
Name: Q4, dtype: int64

In [17]:
df_bachelor_18_21 =  df_18_21[df_18_21['Q4'] == 'Bachelor’s degree']
df_bachelor_18_21.shape

(2288, 355)

In [34]:
df_bachelor_18_21.head(3)

Time from Start to Finish (seconds)     Q1     Q2     Q3  \
11                                7469  18-21    Man  India   
16                                 413  18-21  Woman  India   
20                                 267  18-21    Man  India   

                   Q4       Q5         Q6 Q7_Part_1 Q7_Part_2 Q7_Part_3  ...  \
11  Bachelor’s degree  Student  1-2 years    Python         R       SQL  ...   
16  Bachelor’s degree  Student  1-2 years       NaN         R       SQL  ...   
20  Bachelor’s degree  Student  1-2 years    Python       NaN       NaN  ...   

   Q35_B_Part_2 Q35_B_Part_3 Q35_B_Part_4 Q35_B_Part_5 Q35_B_Part_6  \
11          NaN          NaN          NaN          NaN          NaN   
16          NaN          NaN          NaN          NaN          NaN   
20          NaN          NaN          NaN          NaN          NaN   

   Q35_B_Part_7 Q35_B_Part_8 Q35_B_Part_9 Q35_B_Part_10 Q35_B_OTHER  
11          NaN          NaN          NaN           NaN         NaN  
16          NaN          NaN          NaN          None         NaN  
20          NaN          NaN          NaN           NaN         NaN  

[3 rows x 355 columns]